[PKG-1832] Updates to v1.0.6

schuylermartin45 opened the pull request:
**Changes**
- Updates to v1.0.6

**Issues Encountered**
- N/A

schuylermartin45 committed:
[PKG-1832] Updates to v1.0.6

**Changes**
- Updates to v1.0.6

**Issues Encountered**
- N/A

schuylermartin45 committed:
- Bumps dependency version to latest

schuylermartin45 commented:
`types-requests-feedstock` was not properly linked in `aggregate`. The folder was empty on my local. Will attempt to build that now.

ViridianMelody reviewed (changes requested):
Chained dependencies should be built successfully on a staging channel before review. 
- [Building out dependency chains](https://github.com/anaconda-distribution/perseverance-skills/blob/main/sections/02_Package_building/01_How_tos/Building_out_dependency_chains.md) 
- [PR guidelines](https://github.com/anaconda-distribution/perseverance-skills/blob/main/sections/06_Team_processes/PR_contribution_procedure.md#guidelines)

schuylermartin45 commented:
Turns out @Arishamays1 and @SC426 are both already working on these dependencies and I had no idea, I just assumed (and made an ass of myself) that the packages in `aggregate` were left abandoned.

I'm sure this is not the only time someone duplicated work hunting dependency failures. Just another example how we don't have much in terms of guardrails preventing us from collectively shooting ourselves in the foot.

schuylermartin45 committed:
Fixes dependency on `types-requests`

**Changes**
- Forgot to re-enable this for all platforms
- Realized that this package was already being built by Steve

**Issues Encountered**
- N/A

schuylermartin45 commented:
Prematurely triggered the build. It worked on my local, so I assumed that dependency had been uploaded to the point where Prefect could pull on it. Steve is uploading the dependency now.

SC426 commented on recipe/meta.yaml:
```diff
@@ -12,26 +12,34 @@ source:
 build:
   noarch: python
```
replace with `skip: true  # [py<37]`

ELundby45 replied:
```suggestion
  
```

SC426 commented on recipe/meta.yaml:
```diff
@@ -12,26 +12,34 @@ source:
 requirements:
   host:
     - pip
     - poetry-core >=1.0.0
```
I believe we're still using exact version pinnings in the host section

schuylermartin45 replied:
That check passes in the linter so ¯\_(ツ)_/¯

SC426 replied:
Yeah I wasn't aware that change was made

marcoesters replied:
The recipe standard says the following:

> Version pinnings must have exact versions; comparison operators must not be used. The version number should be the lowest possible. Version pinnings may be skipped for python build packages and packages pinned inside a conda_build_config.yaml file.

`poetry-core` is such an exception because it is a build tool like `pip`, `setuptools`, etc.

There is definitely some refinement needed about which packages need to be pinned and which don't, even though it's not bad being more conservative.

SC426 replied:
How can a noob like me better discern a build tool from a standard package?

marcoesters replied:
That's unfortunately something you just have to read up about. It's not something you can discern by name. To make matters worse, the ecosystem is getting more fractured.

You can check the current linter exceptions here:

https://github.com/anaconda-distribution/anaconda-linter/blob/main/anaconda_linter/lint/check_build_help.py#L32-L51

and here:

https://github.com/anaconda-distribution/anaconda-linter/blob/main/anaconda_linter/lint/check_build_help.py#L119-L123

This is probably not exhaustive. And overall, I'm not sure if there is ever an exhaustive rule that goes beyond "if it's possible for a package update to break an existing build, pin the version".

schuylermartin45 committed:
Incorporates feedback

**Changes**
- Forgot to pull these in from a few other PR comments I've gotten of late

**Issues Encountered**
- N/A

SC426 commented on recipe/meta.yaml:
```diff
@@ -12,26 +12,34 @@ source:
+    - wheel
   run:
+    - py >=1.11.0
     - python >=3.7
```
You can remove all version constraints from `python` in both the host and run section

schuylermartin45 replied:
Already ahead of you on that one

JeanChristopheMorinPerso commented on recipe/meta.yaml:
```diff
@@ -11,27 +11,36 @@ source:
 
 build:
   noarch: python
+  skip: true # [py<37]
```
If I'm not mistaken, Anaconda only provide support for Python 3.8 and newer.

ELundby45 replied:
The team prefers to try and match upstreams requirements regardless of our supported python versions.

schuylermartin45 replied:
Multiple people have told me to start putting this in recipes. I'm guessing that's the reasoning, but nobody has given me a reason as to why we're doing this.

It smells like it needs to be a linter rule to me

marcoesters replied:
The recipe should reflect how the package can be built, not what we currently support. If support were an indicator, we would have to write `skip: true #  [py<38 and py>311]` into every single recipe that builds a python package. And update it with every python version we deprecate and/or add.

The python version we build with are selected via `conda_build_config.yaml`. That way, we separate the way we can build the package according to upstream (`meta.yaml`) and how we actually build the package for Anaconda (`conda_build_config.yaml`).

JeanChristopheMorinPerso commented on recipe/meta.yaml:
```diff
@@ -11,27 +11,36 @@ source:
+    - pytest-cov
     - requests >=2.28.1
+    - types-requests >=2.28.9
+    - pre-commit 3.2.1
```
Upstream i using 2.20 (https://github.com/csernazs/pytest-httpserver/blob/master/pyproject.toml#L74), any reason to pin to 3.2.1 specifically?

schuylermartin45 replied:
This probably also applies to the other comment but I was told that we should pick "stable" dependencies as in "the ones that we support" over necessarily upgrading everything to the ones upstream uses. That is part of our "value add" and I think leans in on the LTS nomenclature.

Not saying it's right or wrong, but that was my understanding about how we operate as a team/company.

schuylermartin45 replied:
And by "ones we support" I mean the package versions we already have upgraded. So I believe 3.2.1 is the latest tagged in our recipe for `pre-commit`, same with `types-toml`. Though I would think minor revisions on a types library should be pretty stable/safe.

marcoesters replied:
I'm not sure I follow that logic. We have `pre-commit 2.20` and it's not an upgrade compared to `3.2.1`. If upstream required `^2.20.0`, we should follow those requirements.

However, I'm not sure all test requirement are really necessary. `pytest-cov` is not needed for a simple `pytest` call and I suspect `pre-commit` may not be needed either to run the tests.

schuylermartin45 replied:
I'm guessing I conflated a build error then and didn't think we had that particular version. I'm guessing there's a `conda` command that lists all of the available versions that would have cleared up my confusion.

As for the second part, the build failed at `import pre-commit` and `import pytest-cov`, respectively without those inclusions

marcoesters replied:
`conda search pre-commit` is what you're looking for.

The import errors probably come from you importing `coverage`. Looking at the upstream feedstock, I would expect you to actually import `pytest_httpserver`.

JeanChristopheMorinPerso commented on recipe/meta.yaml:
```diff
@@ -11,27 +11,36 @@ source:
+    - mypy
     - toml
-    - types-toml >=0.10.8
+    - types-toml >=0.10.8.6
```
Upstream uses `^0.10.2`, which can be translated to `>=0.10.0,<0.11.0` or `0.10.*`. See https://python-poetry.org/docs/dependency-specification/#caret-requirements.

Also, why are some dependency following what's defined in upstream and some don't?

schuylermartin45 committed:
Forgot to delete this line

schuylermartin45 committed:
Adds Marco + Jean's feedback

schuylermartin45 commented on recipe/meta.yaml:
```diff
@@ -27,19 +27,19 @@ requirements:
     - mypy
     - toml
-    - types-toml >=0.10.8.6
+    - packaging
```
The tests would not complete without `packaging` and `execptiongroup` and claimed that `pytest 7.3.1` required them.

marcoesters replied:
`exceptiongroup` shouldn't be needed for python 3.11

marcoesters replied:
This is concerning though. These packages are part of the `pytest` run dependencies and should have been pulled in through pytest.

JeanChristopheMorinPerso replied:
I agree with Marco, it's quite concerning.

JeanChristopheMorinPerso replied:
`conda search -i pytest=7.3.1` says:
```
pytest 7.3.1 py38hca03da5_0
---------------------------
dependencies: 
  - attrs >=19.2.0
  - exceptiongroup >=1.0.0
  - iniconfig
  - packaging
  - pluggy >=0.12,<2.0
  - python >=3.8,<3.9.0a0
  - tomli >=1.0.0

pytest 7.3.1 py39hca03da5_0
---------------------------
dependencies: 
  - attrs >=19.2.0
  - exceptiongroup >=1.0.0
  - iniconfig
  - packaging
  - pluggy >=0.12,<2.0
  - python >=3.9,<3.10.0a0
  - tomli >=1.0.0

pytest 7.3.1 py310hca03da5_0
----------------------------
dependencies: 
  - attrs >=19.2.0
  - exceptiongroup >=1.0.0
  - iniconfig
  - packaging
  - pluggy >=0.12,<2.0
  - python >=3.10,<3.11.0a0
  - tomli >=1.0.0

pytest 7.3.1 py311hca03da5_0
----------------------------
dependencies: 
  - attrs >=19.2.0
  - iniconfig
  - packaging
  - pluggy >=0.12,<2.0
  - python >=3.11,<3.12.0a0
  - tomli >=1.0.0
```

To me pytest is all good and we need to take a look at why you need packaging and exceptiongroup here.

JeanChristopheMorinPerso replied:
I just tested without them and it works. @schuylermartin45 you should be able to remove them. Also, mypy isn't required since you are not running mypy.

schuylermartin45 replied:
This is what I get when I pull `exceptiongroup` and `packaging`
```
Preparing transaction: ...working... done
Verifying transaction: ...working... done
Executing transaction: ...working... done
export PREFIX=/Users/smartin/work/build_ci/pytest-httpserver_1684504037762/_test_env_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_pl
export SRC_DIR=/Users/smartin/work/build_ci/pytest-httpserver_1684504037762/test_tmp
import: 'pytest_httpserver'
import: 'pytest_httpserver'
+ python -m pip check
pytest 7.3.1 requires exceptiongroup, which is not installed.
pytest 7.3.1 requires packaging, which is not installed.
Tests failed for pytest-httpserver-1.0.6-py38hca03da5_0.tar.bz2 - moving package to /Users/smartin/work/build_ci/broken
WARNING:conda_build.build:Tests failed for pytest-httpserver-1.0.6-py38hca03da5_0.tar.bz2 - moving package to /Users/smartin/work/build_ci/broken
TESTS FAILED: pytest-httpserver-1.0.6-py38hca03da5_0.tar.bz2
```

schuylermartin45 replied:
For those of you playing at home, it turns out that my build was using some local and out of date artifacts. Purging the `build-ci` folder fixed my build issues.

marcoesters reviewed (commented):
The test requirements are odd. Technically, they are incorrect, but `pytest-httpserver` is extremely narrow in what they're asking for. If we were to fulfilled them, we'd have to rebuild a lot of packages, which I don't think is worth it.

So, if they pass, we're good? If the requests for changes are changed to approvals, I'm fine with merging as-is, even though the `pytest` package needs to be looked at.

schuylermartin45 committed:
Adds `exceptiongroup` exclusion for `python 3.11`

schuylermartin45 commented:
@JeanChristopheMorinPerso ^ are you comfortable removing the hold?
@ViridianMelody's hold was about a process I wasn't aware, which is good to know, but not an issues anymore once we discovered that I was accidentally trampling over PRs that others were actively working on and have released.

schuylermartin45 committed:
Added @JeanChristopheMorinPerso's feedback

- A local build corruption was causing my build to fail in a way that
  was not reproducible elsewhere

JeanChristopheMorinPerso reviewed (approved):
LGTM. Thanks for making the requested changes!
